### PR TITLE
Implement Hello World in Tonic

### PR DESCRIPTION
**Pull request by conradgrobler**
A basic client and server implementation of the 'hello world' example using Tonic as a gRPC library.

This is the first step towards creating an example of splitting the gRPC server across two proceses as described in #627.

# Checklist

- [ ] Pull request affects core Oak functionality (e.g. runtime, SDK, ABI)
  - [ ] I have written tests that cover the code changes.
  - [ ] I have checked that these tests are run by [Cloudbuild](cloudbuild.yaml)
  - [ ] I have updated [documentation](docs/) accordingly.
  - [ ] I have raised an [issue](https://github.com/project-oak/oak/issues) to
        cover any TODOs and/or unfinished work.
- [x] Pull request includes prototype/experimental work that is under
      construction.

### CODE REVIEW

File: experimental/split_grpc/server/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+// limitations under the License.
+//
+
+use futures_core::Stream;
+use std::pin::Pin;
+use tonic::{transport::Server, Request, Response, Status, Streaming};
+
+pub mod proto {
+    include!("./proto/oak.examples.hello_world.rs");
+}
```
Comment (tiziano88) [inline experimental/split_grpc/server/src/main.rs]: Move this after `use` statements?

File: experimental/split_grpc/server/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+use proto::hello_world_server::{HelloWorld, HelloWorldServer};
+use proto::{HelloRequest, HelloResponse};
+
+#[derive(Debug, Default)]
+pub struct HelloHandler {}
```
Comment (tiziano88) [inline experimental/split_grpc/server/src/main.rs]:
```suggestion
pub struct HelloHandler;
```

Comment (tiziano88) [inline experimental/split_grpc/server/src/main.rs]: (optional) Would `HelloWorldServer` be a more accurate name for this?

File: experimental/split_grpc/server/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        let reply = HelloResponse {
+            reply: format!("Hello {}!", request.into_inner().greeting).into(),
+        };
+
+        Ok(Response::new(reply))
+    }
+
+    // Server streaming response type for the LotsOfReplies method.
+    type LotsOfRepliesStream =
+        Pin<Box<dyn Stream<Item = Result<HelloResponse, Status>> + Send + Sync + 'static>>;
```
Comment (tiziano88) [inline experimental/split_grpc/server/src/main.rs]: Why is this necessary?

File: experimental/split_grpc/client/src/main.rs [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    let mut client = HelloWorldClient::connect("http://[::1]:50051").await?;
+
+    let request = tonic::Request::new(HelloRequest {
+        greeting: "World".into(),
+    });
+
+    let response = client.say_hello(request).await?;
+
+    println!("RESPONSE={:?}", response);
```
Comment (tiziano88) [inline experimental/split_grpc/client/src/main.rs]: May want to switch to https://docs.rs/env_logger/0.7.1/env_logger/ at some point.

Comment (conradgrobler) [reply experimental/split_grpc/client/src/main.rs]: Done.

File: experimental/split_grpc/client/build.rs [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    tonic_build::configure()
+        .build_server(false)
```
[the remark this answers is not in the document]
Comment (conradgrobler) [reply experimental/split_grpc/client/build.rs]: Done.

File: experimental/split_grpc/server/src/main.rs (hunk 1 above)
Comment (conradgrobler) [reply experimental/split_grpc/server/src/main.rs]: Done.

File: experimental/split_grpc/server/src/main.rs (hunk 2 above)
Comment (conradgrobler) [reply experimental/split_grpc/server/src/main.rs]: Done.

File: experimental/split_grpc/server/src/main.rs (hunk 3 above)
Comment (conradgrobler) [reply experimental/split_grpc/server/src/main.rs]: This is mirroring the way Prost generates code, I assume it is to make the function signature more readable.

File: experimental/split_grpc/server/src/main.rs (hunk 2 above)
Comment (conradgrobler) [reply experimental/split_grpc/server/src/main.rs]: Renamed to HelloWorldHandler. There is already a HelloWorldServer that is generated as part of compiling the proto. This provides an implementation of the HelloWorld trait to handle the calls. Alternatively perhaps HelloWorldProvider?

File: experimental/split_grpc/client/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#[macro_use]
+extern crate log;
```
Comment (tiziano88) [inline experimental/split_grpc/client/src/main.rs]: I don't think this is necessary in 2018 edition of Rust. You should be able to just use `log::info!(...)`, or add `use log::info` if you prefer.

Comment (conradgrobler) [reply experimental/split_grpc/client/src/main.rs]: Will include as part of follow-up PR.
